Mathlib/RingTheory/Valuation/Basic.lean

```diff
@@ -889,3 +889,51 @@ scoped[DiscreteValuation] notation "ℕₘ₀" => WithZero (Multiplicative ℕ)
 scoped[DiscreteValuation] notation "ℤₘ₀" => WithZero (Multiplicative ℤ)
 
 end ValuationNotation
+
+section WithZero_Multiplicative_stuff
+
+open Multiplicative
+
+-- in the process of moving these
+theorem WithZero.eq_coe_of_ne_zero {α : Type*} {x : WithZero α}
+    (hx : x ≠ 0) : ∃ (n : α), x = n :=
+  Option.ne_none_iff_exists'.mp hx
+
+theorem WithZero.eq_zero_or_coe {α : Type*} (x : WithZero α) :
+    x = 0 ∨ ∃ (n : α), x = n :=
+  or_iff_not_imp_left.mpr Option.ne_none_iff_exists'.mp
+
+open WithZero
+
+-- this makes `mul_lt_mul_left`, `mul_pos` etc work on `ℤₘ₀`
+instance foo {α : Type*} [Mul α] [Preorder α] [CovariantClass α α (· * ·) (· < ·)]:
+    PosMulStrictMono (WithZero α) where
+  elim := by
+    intro ⟨x, hx⟩ a b (h : a < b)
+    rcases eq_coe_of_ne_zero hx.ne' with ⟨x, rfl⟩
+    dsimp only
+    rcases eq_zero_or_coe a with (rfl | ⟨a, rfl⟩)
+    · rw [mul_zero]
+      rcases eq_coe_of_ne_zero h.ne' with ⟨b, rfl⟩
+      exact WithZero.zero_lt_coe _
+    · have hb : 0 < b := lt_trans (WithZero.zero_lt_coe (ofAdd a)) h
+      rcases eq_coe_of_ne_zero hb.ne' with ⟨b, rfl⟩
+      norm_cast at h ⊢
+      exact mul_lt_mul_left' (α := α) h x
+
+-- This makes `lt_mul_of_le_of_one_lt'` work on `ℤₘ₀`
+instance {α : Type*} [Mul α] [Preorder α] [CovariantClass α α (swap (· * ·)) (· ≤ ·)]:
+    MulPosMono (WithZero α) where
+  elim := by
+    intro ⟨x, hx⟩ a b (h : a ≤ b)
+    dsimp only
+    rcases eq_zero_or_coe x with (rfl | ⟨x, rfl⟩)
```

Comment:
I think it would be better to use `induction x using WithZero.recZeroCoe` here, which avoids the bracket / `rfl` mess, and the need for the new lemma.

Reply:
So you're saying I should use 4 lines to do what I'm doing in one (several times in the proof), because my line is a "mess"? :-)

Reply:
`induction' x with x` is shorter than `rcases eq_zero_or_coe x with (rfl | ⟨x, rfl⟩)` though, right?

Reply:
Oh of course! I was thinking I'd also have to do `subst`. I'll try it this way -- thanks!